src/app/modules/main/wallet_section/all_tokens/io_interface.nim

```diff
@@ -4,7 +4,8 @@ import app/modules/shared_models/currency_amount
 
 type
   SourcesOfTokensModelDataSource* = tuple[
-    getSourcesOfTokensList: proc(): var seq[SupportedSourcesItem]
+    getSourcesOfTokensList: proc(): var seq[SupportedSourcesItem],
+    subscribeToModelUpdates: proc(callback: proc())
```

Comment:
I am not sure about the benefit of this approach, would like to discuss the procs/cons of it. A purpose of having something to "subscribe" has the point if that "unsubscribe" at some point, but here we're subscribing for the events at the app startup (a moment when the model is created) and it's like that for the app's lifetime.

How I see this what's done here:
- module connects to the signals sent from service
- module creates a model over the view
- model subscribes for events to the modulu
- module calls certain model's function via subscribed callback when signal from the service is received, example:
```
self.events.on(SIGNAL_TOKENS_LIST_UPDATED) do(e: Args):
    self.view.setTokenListUpdatedAt(self.controller.getTokenListUpdatedAt())
    for modelUpdatedSubscriber in self.modelUpdatedSubscribers:
      modelUpdatedSubscriber()
``` 

So I don't see benefits, with this additional level of indirection, when all that can be done like this:
- module connects to the signals sent from service
- module creates a model over the view
- module calls certain model's function directly, like `self.view.`via subscribed callback when signal from the service is received, example:
```
self.events.on(SIGNAL_TOKENS_LIST_UPDATED) do(e: Args):
    self.view.setTokenListUpdatedAt(self.controller.getTokenListUpdatedAt())
    self.view.flatTokensModel().modelsUpdated()
``` 

----

Aside from said above, this approach is perfectly fine for setting the delegate interface. `FlatTokenModelDataSource ` should be `FlatTokenModelDelegateInterface` it means, if anyone wants to use `FlatTokensModel` has to provide the delegate which will have implemented all functions that are mentioned in the interface.

----
Changes that would be good to do over `FlatTokenModelDataSource`:
- add token-related delegate interface:
```
TokensDataSourceDlegateInterface* = tuple[
    getFlatTokensList: proc(): var seq[TokenItem],
    getTokenDetails: proc(symbol: string): TokenDetailsItem,
    getTokenPreferences: proc(symbol: string): TokenPreferencesItem
  ]
```
- add UI-related delegate interface:
```
UITokensManagementDlegateInterface* = tuple[
    getCommunityTokenDescription: proc(chainId: int, address: string): string,
    getTokensDetailsLoading: proc(): bool,
    getTokensMarketValuesLoading: proc(): bool
]
```
- remove the following (cause access can be done directly)
```
subscribeToModelUpdates: proc(callback: proc())
subscribeToTokenDetailsUpdates: proc(callback: proc()),
subscribeToTokenMarketValuesUpdates: proc(callback: proc()),
subscribeToTokenPreferencesUpdates: proc(callback: proc()),
subscribeToCurrencyFormatsUpdates: proc(callback: proc())
```

Reply:
@saledjenic I understand your point that this current solution is only an indirection, but what I wanted to achieve here is a way to for model to communicate only with the interface to get data or be updated. 

I can try and think of a better way but, I think calling the model via the module (which is how it is in the current code base) doesn't seem very neat.

about the interface, I dont see a huge advantage in breaking those interfaces as mentioned by you. However if you see `getTokenMarketValuesDataSource` I am passing 2 interfaces to a model and splitting the interfaces based on the source of the data.

Reply:
@Khushboo-dev-cpp I agree it's neat.

What confused me here is the purpose of all this. We have a single place in the app where for example we're using flatTokensModel and it's even defined at the same level as the module that is using it and all signals are there, but we're implementing some interaction unnecessarily.

If we want to follow this approach, then would be good to move all models to the shared location, cause if there is an interface for it, anyone who provides a delegate aligned with the interface defined by the model, can use that model. All functions of the model should be private then and each model should define an "access" interface that the one who is using it should call its methods and a "delegate" interface, that one who is using the model should follow.